File: python/triangle/triangle.py
# coding=utf-8
from __future__ import unicode_literals

__author__ = 'Cedric Zhuang'


class TriangleError(Exception):
    pass


class Triangle(object):
    def __init__(self, a, b, c):
        a, b, c = sorted([a, b, c])
        if a + b <= c:
            raise TriangleError('invalid triangle.')

        if a == c:
            self._kind = "equilateral"
        elif a == b or b == c:
            self._kind = "isosceles"
        else:
            self._kind = "scalene"

    def kind(self):
        return self._kind

File: scala/triangle/src/main/scala/Triangle.scala
/**
  *
  * @author Cedric Zhuang
  *
  */
case class Triangle(a: Int, b: Int, c: Int) {

  import TriangleType._

  lazy val triangleType: TriangleType.Value =
    if (!isLogical) Illogical
    else if (isEquilateral) Equilateral
    else if (isIsosceles) Isosceles
    else Scalene

  lazy val isIsosceles: Boolean = a == b || b == c || a == c
  lazy val isEquilateral: Boolean = a == b && b == c
  lazy val isLogical: Boolean = a + b > c && b + c > a && c + a > b
}

object TriangleType extends Enumeration {
  val Equilateral, Isosceles, Scalene, Illogical = Value
}
